### PR TITLE
Garbage Collect IngestAPI queues

### PR DESCRIPTION
**Pull request by evanxg852000**
Implements the ingest API queue garbage collector

### How was this PR tested?
- unittest

Closes https://github.com/quickwit-oss/quickwit/issues/1229

### CODE REVIEW

File: quickwit-ingest-api/src/queue.rs [hunk 1]
```diff
@@ -264,6 +264,12 @@ impl Queues {
             doc_batch: Some(doc_batch),
         })
     }
+
+    pub fn list_queues(&self) -> crate::Result<ListQueuesResponse> {
+        Ok(ListQueuesResponse {
```
Comment (fulmicoton) [inline quickwit-ingest-api/src/queue.rs:269]: Why result then?

File: quickwit-indexing/src/actors/indexing_service.rs [hunk 2]
```diff
@@ -293,6 +293,23 @@ impl IndexingService {
...
+        &mut self,
+        index_id: String,
+        source_id: String,
+        _ctx: &ActorContext<Self>,
+    ) -> Result<(), IndexingServiceError> {
+        let pipeline_id = IndexingPipelineId {
+            index_id,
+            source_id,
+        };
+        let pipeline_handle_opt = self.pipeline_handles.remove(&pipeline_id);
```
Comment (fulmicoton) [inline quickwit-indexing/src/actors/indexing_service.rs:307]: Considering how little it does, I'd put that logic directly in the handler.

In particular, the passage from Shutdown `message -> (index_id, source_id) positional argument -> IndexingPipelineId` is offering 2 spots where a bug could be hidden from the compiler and the reader view (positional arguments inverted).

To mitigate, you could also have a
`struct Shutdown(IndexingPipelineId);`

File: quickwit-indexing/src/actors/ingest_api_garbage_collector.rs [hunk 3]
```diff
@@ -0,0 +1,146 @@
...
+
+    async fn delete_queue(&self, queue_id: &str) -> Result<(), ActorExitStatus> {
+        // shutdown the pipeline if any
+        self.indexing_service
+            .ask_for_res(ShutdownPipeline {
+                index_id: queue_id.to_string(),
+                source_id: INGEST_API_SOURCE_ID.to_string(),
+            })
+            .await
+            .map_err(|error| anyhow::anyhow!(error))?;
```
Comment (fulmicoton) [inline quickwit-indexing/src/actors/ingest_api_garbage_collector.rs:78]: This actor is unsupervized. It means that if an error is emitted in a handler, the actor will terminated and will not be restarted again.

We may want to be careful about the meaning and the effect of returning an error here.

I suggest we just log an error if that happens.

File: quickwit-indexing/src/actors/ingest_api_garbage_collector.rs [hunk 4]
```diff
@@ -0,0 +1,146 @@
...
+    ) -> Self {
+        Self {
+            metastore,
+            ingest_api_service,
+            indexing_service,
+            counters: IngestApiGarbageCollectorCounters::default(),
+        }
+    }
+
+    async fn delete_queue(&self, queue_id: &str) -> Result<(), ActorExitStatus> {
```
Comment (fulmicoton) [inline quickwit-indexing/src/actors/ingest_api_garbage_collector.rs:70]: Return anyhow::Result here

File: quickwit-indexing/src/actors/ingest_api_garbage_collector.rs [hunk 5]
```diff
@@ -0,0 +1,146 @@
...
+        // shutdown the pipeline if any
+        self.indexing_service
+            .ask_for_res(ShutdownPipeline {
+                index_id: queue_id.to_string(),
+                source_id: INGEST_API_SOURCE_ID.to_string(),
+            })
+            .await
```
Comment (fulmicoton) [inline quickwit-indexing/src/actors/ingest_api_garbage_collector.rs:77]: could be simply `.await`? if we return anyhow::Result.

File: quickwit-indexing/src/actors/ingest_api_garbage_collector.rs [hunk 6]
```diff
@@ -0,0 +1,146 @@
...
+
+#[async_trait]
+impl Handler<Loop> for IngestApiGarbageCollector {
+    type Reply = ();
+
+    async fn handle(&mut self, _: Loop, ctx: &ActorContext<Self>) -> Result<(), ActorExitStatus> {
+        info!("ingest-api-garbage-collect-operation");
+        self.counters.num_passes += 1;
+
+        let queues: HashSet<_> = self
```
Comment (fulmicoton) [inline quickwit-indexing/src/actors/ingest_api_garbage_collector.rs:120]: The type of the element in the HashSet would be helpful here.

File: quickwit-indexing/src/actors/ingest_api_garbage_collector.rs [hunk 7]
```diff
@@ -0,0 +1,146 @@
...
+            .ingest_api_service
+            .ask_for_res(ListQueuesRequest {})
+            .await
+            .map_err(|error| anyhow::anyhow!(error))?
+            .queues
+            .into_iter()
+            .collect();
+
+        let indexes: HashSet<_> = self
```
Comment (fulmicoton) [inline quickwit-indexing/src/actors/ingest_api_garbage_collector.rs:129]: The type of the element in Hashset would be helful

File: quickwit-indexing/src/actors/ingest_api_garbage_collector.rs [hunk 8]
```diff
@@ -0,0 +1,146 @@
...
+            .metastore
+            .list_indexes_metadatas()
+            .await
+            .map_err(|error| anyhow::anyhow!(error))?
+            .into_iter()
+            .map(|index_metadata| index_metadata.index_id)
+            .collect();
+
+        for queue_id in queues.difference(&indexes) {
+            self.delete_queue(queue_id).await?;
```
Comment (fulmicoton) [inline quickwit-indexing/src/actors/ingest_api_garbage_collector.rs:139]: catch the error and log but do no return an error.

**Comment by evanxg852000** [comment]
@fmassot @fulmicoton  Another thing is RocksDB has a default queue that cannot be deleted. if an index is named `default`, the queue will be there all the time. we should guard against this.

**Comment by fmassot** [comment]
As discussed, a restart is always needed on deletion with the CLI with the file-backed metastore.

To avoid the clash with the `default` queue, I would always use `queue_{index-id}` as the name of the queue, but I really don't know if it can cause some issues. If it's too risky, we can just forbid the index name `default` for the 0.3 and remove that constraint later.

File: quickwit-indexing/src/actors/ingest_api_garbage_collector.rs [hunk 9]
```diff
@@ -0,0 +1,235 @@
...
+
+    async fn handle(&mut self, _: Loop, ctx: &ActorContext<Self>) -> Result<(), ActorExitStatus> {
+        info!("ingest-api-garbage-collect-operation");
+        self.counters.num_passes += 1;
+
+        let queues: HashSet<String> = self
+            .ingest_api_service
+            .ask_for_res(ListQueuesRequest {})
+            .await
+            .map_err(|error| anyhow::anyhow!(error))?
```
Comment (fulmicoton) [inline quickwit-indexing/src/actors/ingest_api_garbage_collector.rs:126]: If an error is returned here, the actor will stop running for good today.

**Comment by evanxg852000** [comment]
I think it's safe to internally prefix our queues so I went with `.queue_{index_id}`

**Review by fulmicoton — APPROVED** [review]

**Pull request merged**